source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/lac/petsc_communication_pattern.h>
+
+#ifdef DEAL_II_WITH_PETSC
+
+#  include <deal.II/base/mpi.h>
+
+#  include <deal.II/lac/exceptions.h>
+
+DEAL_II_NAMESPACE_OPEN
+// Shorthand notation for PETSc error codes.
+#  define AssertPETSc(code)                              \
+    do                                                   \
+      {                                                  \
+        PetscErrorCode __ierr = (code);                  \
+        AssertThrow(__ierr == 0, ExcPETScError(__ierr)); \
+      }                                                  \
+    while (0)
```

Comment:
This might be something to move into `exceptions.h` eventually.

Reply:
One wished that PETSc was written in C++ and indicated error conditions by way of an exception. Then we would not have to obfuscate the code for every single PETSc call...

Reply:
Yeah, I know. I'd rather leave this change to a quite big PR with changes to all the PETScWrappers

Reply:
Yes, absolutely!